server/cluster/unsafe_recovery_controller.go

```diff
@@ -166,210 +187,305 @@ func (u *unsafeRecoveryController) reset() {
-		if peer.Role != metapb.PeerRole_Voter && peer.Role != metapb.PeerRole_IncomingVoter {
+		if peer.Role == metapb.PeerRole_Learner {
 			continue
 		}
-		if _, ok := u.failedStores[peer.StoreId]; ok {
-			numFailedVoters += 1
-		} else {
-			numLiveVoters += 1
+		if peer.Role == metapb.PeerRole_Voter || peer.Role == metapb.PeerRole_IncomingVoter {
+			incomingVoters = append(incomingVoters, peer)
+		} else if peer.Role == metapb.PeerRole_Voter || peer.Role == metapb.PeerRole_DemotingVoter {
+			outgoingVoters = append(outgoingVoters, peer)
 		}
 	}
-	return numFailedVoters < numLiveVoters
+
+	return hasQuorum(incomingVoters) && hasQuorum(outgoingVoters)
+
 }
 
-func (u *unsafeRecoveryController) containsFailedPeers(region *metapb.Region) bool {
+func (u *unsafeRecoveryController) getFailedPeers(region *metapb.Region) []*metapb.Peer {
+	var failedPeers []*metapb.Peer
 	for _, peer := range region.Peers {
 		if _, ok := u.failedStores[peer.StoreId]; ok {
-			return true
+			failedPeers = append(failedPeers, peer)
 		}
 	}
-	return false
+	return failedPeers
 }
 
-func keepOneReplica(storeID uint64, region *metapb.Region) {
-	var newPeerList []*metapb.Peer
-	for _, peer := range region.Peers {
-		if peer.StoreId == storeID {
-			if peer.Role != metapb.PeerRole_Voter {
-				peer.Role = metapb.PeerRole_Voter
+var _ btree.Item = &regionItem{}
+
+type regionItem struct {
+	report  *pdpb.PeerReport
+	storeID uint64
+}
+
+func (r *regionItem) Region() *metapb.Region {
+	return r.report.GetRegionState().GetRegion()
+}
+
+func (r *regionItem) IsEpochStale(other *regionItem) bool {
+	re := r.Region().GetRegionEpoch()
+	oe := other.Region().GetRegionEpoch()
+	return re.GetVersion() < oe.GetVersion() || re.GetConfVer() < oe.GetConfVer()
+}
+
+func (r *regionItem) IsStale(origin *regionItem) bool {
+	if r.Region().GetId() != origin.Region().GetId() {
+		panic("should compare peers of same region")
+	}
+
+	// compare region epoch, commit index, term and last index in order
+	if r.IsEpochStale(origin) {
+		return true
+	}
+	re := r.Region().GetRegionEpoch()
+	oe := origin.Region().GetRegionEpoch()
+	if re.GetVersion() == oe.GetVersion() && re.GetConfVer() == oe.GetConfVer() {
+		rs := r.report.GetRaftState()
+		os := origin.report.GetRaftState()
+		if rs.GetHardState().GetCommit() < os.GetHardState().GetCommit() {
+			return true
+		} else if rs.GetHardState().GetCommit() == os.GetHardState().GetCommit() {
+			if rs.GetHardState().GetTerm() < os.GetHardState().GetTerm() {
+				return true
+			} else if rs.GetHardState().GetTerm() == os.GetHardState().GetTerm() {
+				if rs.GetLastIndex() < os.GetLastIndex() {
+					return true
+				}
 			}
-			newPeerList = append(newPeerList, peer)
 		}
 	}
-	region.Peers = newPeerList
+	return false
 }
 
-type peerStorePair struct {
-	peer    *pdpb.PeerReport
-	storeID uint64
+// Less returns true if the region start key is less than the other.
+func (r *regionItem) Less(other btree.Item) bool {
```

Comment:
Is there possible to avoid the duplicated implementation?

Reply:
it's hard as the original `regionTree` accepts `RegionInfo`, but this needs peer report